dali/kernels/imgproc/paste/paste_gpu.h

```diff
@@ -0,0 +1,293 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// THIS WHOLE CODE IS A COPY-PASTE OF multiply-add kernel
+
+#ifndef DALI_KERNELS_IMGPROC_PASTE_PASTE_GPU_H_
+#define DALI_KERNELS_IMGPROC_PASTE_PASTE_GPU_H_
+
+#include <vector>
+#include <unordered_set>
+#include <set>
+#include <map>
+#include <tuple>
+#include "dali/core/span.h"
+#include "dali/core/convert.h"
+#include "dali/core/geom/box.h"
+#include "dali/kernels/imgproc/roi.h"
+#include "dali/kernels/common/block_setup.h"
+#include "dali/kernels/imgproc/paste/paste_gpu_input.h"
+
+namespace dali {
+namespace kernels {
+namespace paste {
+
+template<class InputType, int ndims>
+struct PatchDesc {
+  const InputType *in;
+  ivec<ndims> patch_start, patch_end, in_anchor, in_pitch;
+};
+
+template <class OutputType, class InputType, int ndims>
+struct SampleDescriptorGPU {
+  OutputType *out;
+  int patch_start_idx;
+  ivec<ndims> patch_counts, out_pitch;
+};
+
+// Could not fill this in Setup, as we are not given reference to out then
+template <class OutputType, class InputType, int ndims>
+void FillOutDetails(span<SampleDescriptorGPU<OutputType, InputType, ndims - 1>> out_descs,
+                    const OutListGPU<OutputType, ndims> &out) {
+  int batch_size = out_descs.size();
+
+  for (int i = 0; i < batch_size; i++) {
+    out_descs[i].out = out[i].data;
+  }
+}
+
+/**
+ * Note: Since the operation we perform is channel-agnostic (it is performed in the same
+ * way for every channel), SampleDescriptor assumes, that sample is channel-agnostic. Therefore it
+ * needs to be flattened
+ * @see FlattenChannels
+ */
+template <class OutputType, class InputType, int ndims>
+void CreateSampleDescriptors(
+    vector<SampleDescriptorGPU<OutputType, InputType, ndims - 1>> &out_descs,
+    vector<PatchDesc<InputType, ndims - 1>> &out_patchs,
+    const InListGPU<InputType, ndims> &in,
+    span<paste::MultiPasteSampleInput<ndims - 1>> samples
+) {
+  assert(ndims == 3);
```

Comment:
How does it work with rectangles that are out of output bounds?
For example:
output is 640x480
input starts at 600, 400 and is specified as 100x100.

We should either clip (preferable) or throw, but I don't see it happening here.

Reply:
For now, I added throwing exceptions from the operator if some box is outside in/out bounds.